### ISSUE TITLE
Feedback from 3.0.0 alpha versions

### ISSUE THREAD

**Issue by xoofx** [sharpdx/SharpDX#567]
This issue is to get feedback from the 3.0.0 alpha versions. 

If you have any questions or remarks, about this release, let us know!

Best,
Alexandre

**Comment by QuantumDeveloper** [comment]
Will you provide DX 11.2 binaries for .Net 4.5 via nuget?

**Comment by xoofx** [comment]
@QuantumDeveloper Hm, not sure about which DX11.2 binaries you are referring? D3D11 D3D11.1 D3D11.2 are now all accessible from a single D3D11 assembly.

**Comment by QuantumDeveloper** [comment]
Ah, its good. I remember that earlier D3D11.2 must be built separately as build target DirectX 11_2.

**Comment by xoofx** [comment]
@QuantumDeveloper yes and that was a pain to maintain, build and package. That's one of a big change in 3.0.0

**Comment by QuantumDeveloper** [comment]
Its good change I think)

**Comment by QuantumDeveloper** [comment]
@xoofx Also do you plan to enable SIMD support for math (vectors, matrices)?

**Comment by xoofx** [comment]
> @xoofx Also do you plan to enable SIMD support for math (vectors, matrices)?

As SIMD support is quite orthogonal to SharpDX if you prefer to use `System.Numerics.Vectors` instead of `SharpDX.Mathematics`, it is up to your project. There is currently no plan to migrate `SharpDX.Mathematics` to `System.Numerics.Vectors`.

**Comment by xanather** [comment]
Thanks for your hard work on this release. I'm really happy about the basic interop types. Trying to develop a basic engine/framework ontop of SharpDX while trying to encapsulate SharpDX data types was a strange thing.

**Comment by emzic** [comment]
Thanks for your continuous work on SharpDX!

Will the Toolkit be updated to compile with the latest changes?

**Comment by xoofx** [comment]
> Will the Toolkit be updated to compile with the latest changes? 

Unfortunately, no, the Toolkit is deprecated and no longer supported, as I'm focusing on [Paradox](http://paradox3d.net/) now. But if someone wants to fix it, pull requests are welcome.

**Comment by stgeorge** [comment]
Wanted to say thank you to @xoofx for his continued work on SharpDX which is a great boon for C# and Direct3D hackers.

**Comment by Gavin-Williams** [comment]
Maths classes were meant to be in Mathematics, but the alpha has moved them all back to SharpDX, under the old 3.0.0 they were already in SharpDX.Mathematics.

**Comment by xoofx** [comment]
> Maths classes were meant to be in Mathematics, but the alpha has moved them all back to SharpDX, under the old 3.0.0 they were already in SharpDX.Mathematics.

Yeah, sorry, It was breaking too many samples and wanted to avoid this burden for existing users as well, so I reverted it back to the root namespace. If you look at existing code base, when there was a `using SharpDX;` it was only used to include Mathematics and just the `Utilities` class, so it is fine to let it as it is today. As the SharpDX namespace is a bit less cluttered it should be fine. Also the `SharpDX.Mathematics` assembly is optional, so it doesn't force to use it.

**Comment by Gavin-Williams** [comment]
ok.

Also want to say that i really like the simplicity of having desktop & store versions. It must be a lot easier to maintain. And thanks for the DX12 preview. I probably won't look at it myself for a few months, because my current project is undergoing enough twists and turns as it is, but looking forward to eventually taking a peek. Thanks for all your hard work.

**Comment by xanather** [comment]
I just converted to using the alpha binaries.

The alpha version seems to be throwing and catching many System.Exception's internally, I think its slowing down my application. Any standard Direct3D11 call is causing it, i.e. SwapChain.ResizeBuffers, and even the SharpDX.Direct3D11.Device constructor.

I am using the Release/Retail compiled SharpDX alpha assemblies always. When compiling my project under Debug it lags, only while the debugger is attached, otherwise it is fine (i.e. running the Debug executable outside of visual studio is OK). Running the project in release mode is also fine.

Visual studio only detects System.Exception's while running in Debug mode for my project.

Its really odd, I'm still not 100% certain it has something to do with SharpDX. I have however reverted back to the 2.6.3 assemblies to test and the problem is fixed (no System.Exception's detected anywhere).

**Comment by xoofx** [comment]
@xanather I have clone the Samples repo and downloaded the zip with 3.0.0-alpha01, and samples are working without any exceptions, so it must be something wrong in your setup. If you still have an issue, open a new issue so that we will discuss specifically from there.

**Comment by xanather** [comment]
Thanks for testing that, given the samples don't show anything I will investigate more on my side before deciding if I should setup an issue.

**Comment by Gavin-Williams** [comment]
I don't know where this issue lies, but please take a look, as i can't pin it down, and it's starting to look like the problem doesn't lie with the xaml rendering call.
https://social.msdn.microsoft.com/Forums/windowsapps/en-US/34848738-61ec-4af5-b46e-b0b6fda6c58a/cant-prevent-multiple-threads-from-calling-directx?forum=winappswithcsharp

**Comment by zachsaw** [comment]
I just read on the SharpDX news page that Direct3D10 APIs have been removed from v3.0. Is there any specific reason as to why? My media player relies on D3D9, 10 and 11 and in the near future I plan to support D3D12 as well. How would I go about doing this with SharpDX v3.0? Would it be possible to use v3.0 alongside v2.6.3? I can't drop D3D10 support as there are old cards out there that is still perfectly capable of playing video that don't have D3D11 and requires D3D10's exclusive mode for 10/16-bit output.

**Comment by xoofx** [comment]
@zachsaw, HW supported by d3d10 is also supported by d3d11 (this is called HW feature level, a reason why you can use d3d11 api on a mobile phone with a 9.3 GPU) so afaik There is no reason to support d3d10 (only reason would be for d3d10/d2d interop on older win7 OS)

**Comment by zachsaw** [comment]
@xoofx, Ah I see, so you'd just use d3d11 API and request for feature_level_10? What about the Font APIs of d3d10?

**Comment by QuantumDeveloper** [comment]
@xoofx Does effects framework also available for DX12 in sharpdx?

**Comment by xanather** [comment]
The effects framework was depreciated in Direct3D11. I assume it doesn't exist in Direct3D12 meaning SharpDX can't provide it.

**Comment by QuantumDeveloper** [comment]
@xanather really? then what effects do in direct3d11_2 ddls of sharpdx?

**Comment by morpheusxx** [comment]
@xoofx 
On the SharpDX homepage is written:
"All Mathematics (Vector3, Vector4...) have been moved to a dedicated assembly. The reason behind this change is to put SharpDX on a diet, where higher level API (like Paradox, MonoGame...) are no longer forced to include this API. In order to support this, all SharpDX assemblies are now using interop types (RawVector3, RawVector4 from SharpDX.Mathematics.Interop namespace)"

I really want to continue using the classes from SharpDX.Mathematics assembly, because our SkinEngine makes havy use of calculation with RectangleF, SizeF, Matrix etc. for layouting.

How it is supposed to use SharpDX3 Raw* types in combination with the classes from Mathematics assembly? Are there any conversion routines available? 

I already tried to workaround with creation of extension methods for RectangleF, but once I get to D2D custom effects which require RawMatrix while all calculations do use Matrix, the situation is very bad.

Any guidance for a v3 migration of our project is very appreciated!

(I hope to solve compatibility issues with Direct2D1 and Windows 10)

**Comment by xanather** [comment]
The SharpDX math/geometry types should be able to implicitly convert to its respective raw type:

https://github.com/sharpdx/SharpDX/blob/master/Source/SharpDX.Mathematics/Vector3.cs#L1808
https://github.com/sharpdx/SharpDX/blob/master/Source/SharpDX.Mathematics/Color.cs#L1058
https://github.com/sharpdx/SharpDX/blob/master/Source/SharpDX.Mathematics/Point.cs#L144

For example, just provide a Point type if it asks for a RawPoint (or any other type for that matter).

**Comment by xoofx** [comment]
@morpheusxx for D2D effects, the latest 3.0 version is not yet working, so it will change slightly the declaration of the custom attribute on the field, but than after it will work directly without changing the types for the field (you will be able to keep Matrix/Vector...etc.)

**Comment by PierreCompute** [comment]
Thanks for the continued versions of SharpDX!
I've been using SharpDX 2.x to enable DirectCompute in .NET.  I understand the toolkit will not move to 3.0, but really liked the GetData and SetData functions in the toolkit for structured,raw,and constant buffers.  I see that it is very easy to create initialized buffers in SharpDX 3.0 by passing in a typed array (or structure for constant buffers) into Buffer.Create().  Are there plans to extend that ease with generic GetData and SetData functions?  
Also I notice there is support for vector234, and half234.  Is it possible to add support for double234?

**Comment by chedburgh** [comment]
Also just tried version 3.0.0.0, and like zachsaw found Direct3D10 missing. I am writing a Direct2D accelerated control for a WPF application targeted at Windows 7 – 10. 

Couple of questions, forgive me if they are simple, but I am not familiar with the interop between various DirectX releases.

@xoofx Can I support Direct2d using Direct3D11 on Windows 7 – 10? Or is Direct3D10 a must for Windows 7? Will it be possible for me to move to 3.0.0.0 and continue supporting Windows 7 – 10 with my Direct2D control?

Thanks for your time.

**Comment by xanather** [comment]
@chedburgh Direct3D11 works on vista and higher, i suppose that's why not many use Direct3D10 anymore since Direct3D10 also works on vista only or higher. Same requirements just Direct3D11 is newer and more evolved. Most GPU's that support Direct3D10 also support Direct3D11. Not sure about what Direct2D needs to use underneath though.

**Comment by xoofx** [comment]
@chedburgh Direct2D should work with Direct3D11 with [Windows 7 platform update](https://support.microsoft.com/en-us/kb/2670838#) installed. But I'm not sure this platform update is installed automatically, so maybe I will consider to re-introduce Direct3D10 assembly.

**Comment by chedburgh** [comment]
Thanks both for the answers.

@xanather I am not sure either. I currently have Direct2D working with the Direct3D10/D3DImage fine for my needs (finally!), and I'm not sure on the requirements to move it to Direct3D11. Something for me to investigate later.

@xoofx Thanks for considering it. I'll keep an eye on the commits/news/twitter feed and move across to help testing 3.0.0.0 if I see it reintroduced.

**Comment by bdachev** [comment]
Hi Guys, I have been playing around with Direct3D12 and have some problems to report if someone is interested.

**Comment by xoofx** [comment]
@bdachev sure, open a dedicated issue for this

**Comment by bdachev** [comment]
OK. I will open dedicated issue for each problem I find. In general I started doing my tests before SharpDX 3.0 alpha2 was published so I had to manually patch the SharpDX source code to compile against newer Windows 10 SDK. I also use here Visual Studio 2015 RC and probably my patches won't work directly for you.

**Comment by bdachev** [comment]
Just for the record after pulling the latest from master and compiling it here it seems the problems I was having are already resolved. There were some alignment issues due to pointers in structures but it is OK now for both x86 and x64 configurations.

**Comment by axefrog** [comment]
I've been experimenting with using SharpDX 3 in a CoreCLR project and I've discovered that there is no option in a Visual Studio 2015 CoreCLR project to enable native code debugging. Are there any other ways around this that you're aware of? Or are we a bit dead in the water in this regard?

**Comment by xanather** [comment]
@axefrog Are you using DNX? Maybe its something to post about here https://github.com/aspnet/DNX as it's not really related to SharpDX.

**Comment by axefrog** [comment]
@xanather I'm using VS2015 as the issue relates to me not being able to see DX debug messages in the debug output panel. I've had a huge amount of trouble finding anyone who can provide a canonical answer on this, but I am beginning to suspect that you're probably right; it may simply be too early in CoreCLR's development cycle.

**Comment by axefrog** [comment]
Just thought I'd leave this here now that I finally have an answer. If you're trying to use SharpDX 3 in one of the new VS2015 .Net Core projects (with the project.json file, dnx and all of that stuff), you can still debug and get native Direct3D debug messages. This StackOverflow question has the details. http://stackoverflow.com/questions/31558596/no-native-code-debugging-in-coreclr-console-application-projects-in-vs2015

**Comment by RobyDX** [comment]
Hi, I'm using SharpDx D3D12 to prepare a set of tutorials (I'm actually experimenting it to learn as much as possible). 
Except for one bug (I've added a pull request for ResorceDescription) it works great. 

Just missing some utilities (like loading resource from file, I'm writing a method to load dds)  and DescriptorRange array inside DescriptorParameter. For now I'm doing this way

```javascript
DescriptorRange[] ranges = ...
.............
desc= new RootDescriptorTable()
    {
    DescriptorRangeCount = ranges.Length,
    PDescriptorRanges = Marshal.UnsafeAddrOfPinnedArrayElement(ranges, 0)
}
```

Any prevision for next release? 
Thx

**Comment by xoofx** [comment]
@RobyDX I have been busy recently, hopefully I will have a look to upgrade Direct3D12 next week.

**Comment by RobyDX** [comment]
That's will be great. If you need some help, ask me. 
I'm really interested in D3D12 for .Net

**Comment by manofstick** [comment]
@xoofx 

Thanks for all your hard work!

I'm using Direct2D (from F#) and just did a trial run update to 3.0.0 (beta1). (

First up, all the implict conversions don't work in F#, so that's a little bit painful, but can be resolved with some f# extension methods such as (some random examples, covering some different cases):

    [<AutoOpen>]
    module ImplicitCoversion =
        let inline (!>) (x:^a) : ^b = ((^a or ^b) : (static member op_Implicit : ^a -> ^b) x) 

    type SharpDX.Direct2D1.RenderTarget with
        member inline this.DrawLine(v0:Vector2, v1:Vector2, brush) =
            this.DrawLine (!> v0, !> v1, brush)

        member inline this.Transform
            with set (value:Matrix3x2) = this.Transform <- !> value

        member inline this.Clear (color:Color4) =
            this.Clear (Nullable (!> color))

    type SharpDX.Direct2D1.SolidColorBrush with
        static member inline Create(renderTarget, color:Color4) = new SolidColorBrush(renderTarget, !> color)

Which seem to work pretty well. Maybe I'll create an exhaustive set at some stage so assist any others using f# - would you want something like that "officially" in your repo (i.e. probably an extra assembly given the f#ness)

But what I couldn't "simply" solve was SimplifiedGeometrySink.AddLines because it takes an array.  Just wondering if you have an magic for casting between array types? (and if so, maybe an extension method could be added for SimplifiedGeometrySink in the SharpDX.Mathematics library?)

**Comment by ulao** [comment]
I have been working on a force feed back example for a few days now only to find I'm hitting a wall. I just saw this thread and wondered if its why I can not get it working. When I create the effect I get parameter incorrect. I can provide details if needed.
